mempool/reactor.go

```diff
@@ -134,31 +161,87 @@ func (memR *Reactor) AddPeer(peer p2p.Peer) {
 	}
 }
 
+func (memR *Reactor) RemovePeer(peer p2p.Peer, _ any) {
+	if memR.router != nil {
+		// Remove all routes with peer as source or target and immediately
+		// adjust redundancy.
+		memR.router.resetRoutes(peer.ID())
+		memR.redundancyControl.triggerAdjustment(memR)
+		memR.mempool.metrics.DisabledRoutes.Set(float64(memR.router.numRoutes()))
+	}
+}
+
 // Receive implements Reactor.
 // It adds any received transactions to the mempool.
 func (memR *Reactor) Receive(e p2p.Envelope) {
-	memR.Logger.Debug("Receive", "src", e.Src, "chId", e.ChannelID, "msg", e.Message)
-	switch msg := e.Message.(type) {
-	case *protomem.Txs:
-		if memR.WaitSync() {
-			memR.Logger.Debug("Ignored message received while syncing", "msg", msg)
-			return
-		}
+	if memR.WaitSync() {
+		memR.Logger.Debug("Ignore message received while syncing", "src", e.Src, "chId", e.ChannelID, "msg", e.Message)
+		return
+	}
 
-		protoTxs := msg.GetTxs()
-		if len(protoTxs) == 0 {
-			memR.Logger.Error("Received empty Txs message from peer", "src", e.Src)
-			return
+	senderID := e.Src.ID()
+
+	switch e.ChannelID {
+	case MempoolControlChannel:
+		switch msg := e.Message.(type) {
+		case *protomem.HaveTx:
+			txKey := types.TxKey(msg.GetTxKey())
+			if len(txKey) == 0 {
+				memR.Logger.Error("Received empty HaveTx message from peer", "src", e.Src.ID())
```

Comment:
IMHO, this should also lead to us disconnecting from this peer, but that's a point for another PR, I guess.

Reply:
That's correct. For another PR.